File: src/js/core/src/Style/Meta.ts
import type { StyleDefinitionMeta } from '../StyleDefinition.js';
import { MetaLicense } from './MetaLicense.js';
import { MetaCreator } from './MetaCreator.js';
import { MetaSource } from './MetaSource.js';

/**
 * Lazily-constructed view over a style definition's `meta` block, exposing
 * the license, creator, and source descriptors.
 */
export class Meta {
  #data: StyleDefinitionMeta;
  #license?: MetaLicense;
  #creator?: MetaCreator;
  #source?: MetaSource;

  constructor(data: StyleDefinitionMeta) {
    this.#data = data;
  }

  /**
   * Returns the license descriptor, defaulting to an empty object when the
   * style definition omits the field.
   */
  license(): MetaLicense {
    this.#license ??= new MetaLicense(this.#data.license ?? {});

    return this.#license;
  }

  /**
   * Returns the creator descriptor, defaulting to an empty object when the
   * style definition omits the field.
   */
  creator(): MetaCreator {
    this.#creator ??= new MetaCreator(this.#data.creator ?? {});

    return this.#creator;
  }

  /**
   * Returns the source descriptor, defaulting to an empty object when the
   * style definition omits the field.
   */
  source(): MetaSource {
    this.#source ??= new MetaSource(this.#data.source ?? {});

    return this.#source;
  }
}

File: src/dart/core/lib/src/style/meta.dart
/// Read-only view over a style definition's `meta` block.
library;

import 'meta_creator.dart';
import 'meta_license.dart';
import 'meta_source.dart';

/// The metadata of a style, exposing the license, creator, and source
/// descriptors. Each descriptor defaults to an empty block when the style
/// definition omits the field, so callers never need a null check.
class Meta {
  final Map<String, Object?> _data;
  MetaLicense? _license;
  MetaCreator? _creator;
  MetaSource? _source;

  Meta(Map<String, Object?> data) : _data = data;

  /// Returns the license descriptor, lazily constructed on first access and
  /// defaulting to an empty block when the style definition omits the field.
  MetaLicense license() => _license ??=
      MetaLicense(_data['license'] as Map<String, Object?>? ?? const {});

  /// Returns the creator descriptor, lazily constructed on first access and
  /// defaulting to an empty block when the style definition omits the field.
  MetaCreator creator() => _creator ??=
      MetaCreator(_data['creator'] as Map<String, Object?>? ?? const {});

  /// Returns the source descriptor, lazily constructed on first access and
  /// defaulting to an empty block when the style definition omits the field.
  MetaSource source() => _source ??=
      MetaSource(_data['source'] as Map<String, Object?>? ?? const {});
}
